### PR TITLE
getistio fetch: add name flag in fetch

### PR DESCRIPTION
**Pull request by ZhiHanZ**
resolve https://github.com/tetratelabs/getistio/issues/19

### CODE REVIEW

**Review by mathetake — COMMENTED** [review]
shall we add `NAME` column in `getistio list` command for convenience?

File: src/istioctl/istioctl.go [hunk 1]
```diff
@@ -203,17 +203,24 @@ func ExecWithWriters(homeDir string, args []string, stdout, stderr io.Writer) er
...
 	return d, fetch(homeDir, d, ms)
 }
 
-func processFetchParams(version, flavor string, flavorVersion int,
+func processFetchParams(name, version, flavor string, flavorVersion int,
 	ms *api.Manifest) (*api.IstioDistribution, error) {
+	if len(name) != 0 {
+		d, err := api.IstioDistributionFromString(name)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse given name to %s istio distribution", name)
```
Comment (mathetake) [inline src/istioctl/istioctl.go:220]:
```suggestion
			return nil, fmt.Errorf("cannot parse given name %s  to istio distribution", name)
```

Comment (mathetake) [reply src/istioctl/istioctl.go:220]: Could you do the same fix for https://github.com/tetratelabs/getistio/blob/ddc6cb8c3305eae74732fa7500fe5e64ed9bc59d/cmd/switch.go#L91 ?

**Review by mathetake — CHANGES_REQUESTED** [review]
shall we put `istioctl.processFetchParams` in `fetch.go` and make `istioct.Fetch` accepting `*api.IstioDistribution`  to be better aligned with switch cmd?

File: cmd/fetch.go [hunk 2]
```diff
@@ -97,8 +110,85 @@ For more information, please refer to "getistio list --help" command.
...
-	flags.StringVarP(&flagFlavor, "flavor", "", "", "Flavor of istioctl, e.g. \"--flavor tetrate\" or --flavor tetratefips\" or --flavor istio\"")
-	flags.IntVarP(&flagFlavorVersion, "flavor-version", "", -1, "Version of the flavor, e.g. \"--version 1\"")
+	flags.StringVarP(&flag.Name, "name", "", "", "Name of distribution, e.g. 1.9.0-istio-v0")
+	flags.StringVarP(&flag.Version, "version", "", "", "Version of istioctl e.g. \"--version 1.7.4\". When --name flag is set, this will not be used.")
+	flags.StringVarP(&flag.Flavor, "flavor", "", "", "Flavor of istioctl, e.g. \"--flavor tetrate\" or --flavor tetratefips\" or --flavor istio\". When --name flag is set, this will not be used.")
+	flags.Int64VarP(&flag.FlavorVersion, "flavor-version", "", -1, "Version of the flavor, e.g. \"--version 1\". When --name flag is set, this will not be used.")
 	return cmd
 }
+
+func processFetchParams(flags *FetchFlags,
```
Comment (mathetake) [inline cmd/fetch.go:120]: Let's name `fetchParse` following `switchParse`

**Comment by ZhiHanZ** [comment]
> shall we put `istioctl.processFetchParams` in `cmd/fetch.go` and make `istioct.Fetch` accepting `*api.IstioDistribution` to be better aligned with switch cmd?



> shall we add `NAME` column in `getistio list` command for convenience?

https://github.com/tetratelabs/getistio/issues/28

**Comment by ZhiHanZ** [comment]
> shall we add `NAME` column in `getistio list` command for convenience?
Created an issue on it

https://github.com/tetratelabs/getistio/issues/28

File: src/istioctl/istioctl.go [hunk 3]
```diff
@@ -203,84 +201,10 @@ func ExecWithWriters(homeDir string, args []string, stdout, stderr io.Writer) er
 	return cmd.Run()
 }
 
-func Fetch(homeDir string, version, flavor string, flavorVersion int, ms *api.Manifest) (*api.IstioDistribution, error) {
-	d, err := processFetchParams(version, flavor, flavorVersion, ms)
-	if err != nil {
-		return nil, err
-	}
-
+func Fetch(homeDir string, d *api.IstioDistribution, ms *api.Manifest) (*api.IstioDistribution, error) {
```
Comment (mathetake) [inline src/istioctl/istioctl.go:204]: I think we no longer need to return distro

```suggestion
func Fetch(homeDir string, target *api.IstioDistribution, ms *api.Manifest) error {
```

File: src/istioctl/istioctl.go [hunk 4]
```diff
@@ -203,84 +201,10 @@ func ExecWithWriters(homeDir string, args []string, stdout, stderr io.Writer) er
...
 	return d, fetch(homeDir, d, ms)
```
Comment (mathetake) [inline src/istioctl/istioctl.go:205]: put `fetch`'s logic directly here. I don't think we need to have separate `Fetch` and `fetch`.

**Review by mathetake — CHANGES_REQUESTED** [review]

File: src/istioctl/istioctl_test.go [hunk 5]
```diff
@@ -378,53 +378,6 @@ func TestFetch(t *testing.T) {
...
-			},
-			{
-				Version:       "1.7.5",
-				Flavor:        api.IstioDistributionFlavorTetrate,
-				FlavorVersion: 0,
-			},
-		},
-	}
-
-	t.Run("not-supported", func(t *testing.T) {
```
Comment (mathetake) [inline src/istioctl/istioctl_test.go:407]: why did you delete this case?

File: src/istioctl/istioctl_test.go [hunk 6]
```diff
@@ -378,53 +378,6 @@ func TestFetch(t *testing.T) {
...
-			{Version: "1000.7.4", Flavor: api.IstioDistributionFlavorTetrate},
-			{Version: "1.7.5", Flavor: api.IstioDistributionFlavorTetrateFIPS},
-			{Version: "1.7.5", Flavor: api.IstioDistributionFlavorTetrate, FlavorVersion: 1},
-		} {
-			_, err = Fetch(dir, c, ms)
-			require.Error(t, err)
-		}
-	})
-
-	t.Run("supported", func(t *testing.T) {
```
Comment (mathetake) [inline src/istioctl/istioctl_test.go:418]: same here

File: cmd/fetch.go [hunk 7]
```diff
@@ -16,19 +16,25 @@ package cmd
...
+	"github.com/Masterminds/semver"
 	"github.com/spf13/cobra"
 
+	"github.com/tetratelabs/getistio/api"
 	"github.com/tetratelabs/getistio/src/istioctl"
 	"github.com/tetratelabs/getistio/src/manifest"
 	"github.com/tetratelabs/getistio/src/util/logger"
 )
 
+type FetchFlags struct {
```
Comment (mathetake) [inline cmd/fetch.go:30]: no need to export

File: cmd/fetch.go [hunk 8]
```diff
@@ -16,19 +16,25 @@ package cmd
...
+	Name, Version, Flavor string
+	FlavorVersion         int64
```
Comment (mathetake) [inline cmd/fetch.go:32]: same here

File: cmd/fetch.go [hunk 9]
```diff
@@ -16,19 +16,25 @@ package cmd
...
+}
+
 func newFetchCmd(homedir string) *cobra.Command {
 	var (
```
Comment (mathetake) [inline cmd/fetch.go:36]: can be simplified: `var flag FetchFlags`

File: cmd/fetch.go [hunk 10]
```diff
@@ -97,8 +110,85 @@ For more information, please refer to "getistio list --help" command.
...
+func fetchParams(flags *FetchFlags,
+	ms *api.Manifest) (*api.IstioDistribution, error) {
+	if len(flags.Name) != 0 {
+		d, err := api.IstioDistributionFromString(flags.Name)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse given name %s to istio distribution", flags.Name)
+		}
+		return d, nil
+	}
+	if flags.Flavor != api.IstioDistributionFlavorTetrate && flags.Flavor != api.IstioDistributionFlavorTetrateFIPS && flags.Flavor != api.IstioDistributionFlavorIstio {
```
Comment (mathetake) [inline cmd/fetch.go:129]: shall we have new line somewhere in the middle of this line since it looks too long

File: cmd/fetch.go [hunk 11]
```diff
@@ -97,8 +110,85 @@ For more information, please refer to "getistio list --help" command.
 
 	flags := cmd.Flags()
 	flags.SortFlags = false
-	flags.StringVarP(&flagVersion, "version", "", "", "Version of istioctl e.g. \"--version 1.7.4\"")
-	flags.StringVarP(&flagFlavor, "flavor", "", "", "Flavor of istioctl, e.g. \"--flavor tetrate\" or --flavor tetratefips\" or --flavor istio\"")
-	flags.IntVarP(&flagFlavorVersion, "flavor-version", "", -1, "Version of the flavor, e.g. \"--version 1\"")
+	flags.StringVarP(&flag.Name, "name", "", "", "Name of distribution, e.g. 1.9.0-istio-v0")
+	flags.StringVarP(&flag.Version, "version", "", "", "Version of istioctl e.g. \"--version 1.7.4\". When --name flag is set, this will not be used.")
+	flags.StringVarP(&flag.Flavor, "flavor", "", "", "Flavor of istioctl, e.g. \"--flavor tetrate\" or --flavor tetratefips\" or --flavor istio\". When --name flag is set, this will not be used.")
+	flags.Int64VarP(&flag.FlavorVersion, "flavor-version", "", -1, "Version of the flavor, e.g. \"--version 1\". When --name flag is set, this will not be used.")
```
Comment (mathetake) [inline cmd/fetch.go:116]: same here: let's have new line somewhere in the middle

**Review by mathetake — APPROVED** [review]
LGTM
